### DISCUSSION TITLE
Garbled text in Pdf in Emacs 29.4

### DISCUSSION THREAD

**Discussion by deccanqueen88** [vedang/pdf-tools#291] [Q&A]
OS: Pop!_OS 22.04 LTS

Hi everyone, this is the first time I'm using github, and will be grateful for the community's help. Firstly, thank you to the developers for this wonderful tool. I am a Humanities researcher and have been primarily using emacs for academic writing.
I was using pdf-tools with GNU Emacs 27.1 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.33, cairo version 1.16.0) for about a year or so now, and it's been great. I have been trying to upgrade to GNU Emacs 29.4 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.42, cairo version 1.18.0), but the main thing holding me back is that I am just not able to get pdf-tools working.  I was able to get it installed with M-x list-packages, install pdf-tools from MELPA, but when I open a pdf, all I see is this garbled text. 
![Screenshot from 2024-08-10 00-04-36](https://github.com/user-attachments/assets/8cd68b2a-f3ec-447a-817c-4a64d7b6909a)

Also, while I succeeded in installing pdf-tools through M-x list-packages, when I start up emacs, it is not doing it automatically and producing an error. 
This is what I have in my config file: 

 #+begin_src emacs-lisp

(use-package pdf-tools
:ensure t
:config
(custom-set-variables
        '(pdf-tools-handle-upgrades nil)) 
     (setq pdf-info-epdfinfo-program "/usr/local/bin/epdfinfo"))
(pdf-tools-install))

(use-package org-pdfview
:ensure t)
#+end_src>

On start up, this is the message I get : 

Loading /home/brani/.emacs.d/config.el (source)...
.emacs.d/config.el: Warning: (lambda nil \...) quoted with ' rather than with #'
Loading /home/brani/.emacs.d/config.el (source)...done
Loaded /home/brani/.emacs.d/config.el
Need to (re)build the epdfinfo program, do it now ? (y or n) y
For information about GNU Emacs and the GNU system, type C-h C-a.
Comint exited abnormally with code 1
error in process sentinel: Building the PDF Tools server failed [2 times]
Updating buffer list...
Formats have changed, recompiling...done
Updating buffer list...done
Commands: m, u, t, RET, g, k, S, D, Q; q to quit; h for help
Mark set

I don't particularly understand code (lisp or any other) and have been using emacs through the help of the many discussions online on problems I face, but I have just not been able to figure this out. Also, my main motivation for upgrading emacs version is so that I can use org-cite, which is lacking in the 27.1 version. I'll be grateful if you can help me get this sorted.

Thank   you.

**Comment by vedang** [comment]
As you can see from your messages buffer, building `pdf-tools` server (`epdfinfo`) is failing. Based on your configuration, maybe you have installed `pdf-tools` from some distribution package manager? This is not necessary. 

Change your configuration to:
```
(use-package pdf-tools
 :ensure t
 :config (pdf-tools-install))
```

and try again. 

If/When it fails, share the contents of the compilation buffer, which will have details of what has failed. (To find the compilation buffer, `M-x ibuffer`, search for `compilation`, `RET`).

**Comment by deccanqueen88** [comment]
Hi Vedang, thanks for getting back to me. I have changed the config file as you suggested. I then restarted emacs and had to run `M-x pdf-tools-install` and this is the Compilation output I am getting:

```
-*- mode: compilation; default-directory: "~/.emacs.d/elpa/pdf-tools-20240429.407/build/server/" -*-
Comint started at Sat Aug 10 12:07:51

/home/brani/.emacs.d/elpa/pdf-tools-20240429.407/build/server/autobuild -i /home/brani/.emacs.d/elpa/pdf-tools-20240429.407/
Failed to recognize this system, trying to continue.
---------------------------
 Configuring and compiling 
---------------------------
./configure -q --bindir=/home/brani/.emacs.d/elpa/pdf-tools-20240429.407/ && make clean && make -s
configure: error: Package requirements (poppler) were not met:

Package 'poppler' not found

Consider adjusting the PKG_CONFIG_PATH environment variable if you
installed software in a non-standard prefix.

Alternatively, you may set the environment variables poppler_CFLAGS
and poppler_LIBS to avoid the need to call pkg-config.
See the pkg-config man page for more details.
===========================
     Build failed.  ;o(    
===========================
Note: maybe try the '-d' option.

Comint exited abnormally with code 1 at Sat Aug 10 12:07:52
```

**Comment by vedang** [comment]
Go to `/home/brani/.emacs.d/elpa/pdf-tools-20240429.407/` in the terminal, and run `./server/autobuild`. This should install the missing dependencies ("poppler") and then build pdf-tools.

Then restart Emacs. Use the new config that I shared here, you don't need the older config.

If the `./server/autobuild` step fails, share the output of that step

**Reply by deccanqueen88** [reply]
Hi, I have done this, and the build seems to have been successful. 

![Screenshot from 2024-08-10 15-39-38](https://github.com/user-attachments/assets/f3f810f4-ff24-4166-b95e-f1d6c3a7bc99)

But when I start emacs again, and open a pdf, it is still the same garbled text. 
When I try to run M-x pdf-tools-install in emacs now, I get the same error (despite the autobuild in the terminal saying that the build was successful):

-*- mode: compilation; default-directory: "~/.emacs.d/elpa/pdf-tools-20240429.407/build/server/" -*-
Comint started at Sat Aug 10 15:43:07

/home/brani/.emacs.d/elpa/pdf-tools-20240429.407/build/server/autobuild -i /home/brani/.emacs.d/elpa/pdf-tools-20240429.407/
Failed to recognize this system, trying to continue.
---------------------------
 Configuring and compiling 
---------------------------
./configure -q --bindir=/home/brani/.emacs.d/elpa/pdf-tools-20240429.407/ && make clean && make -s
configure: error: Package requirements (poppler) were not met:

Package 'poppler' not found

Consider adjusting the PKG_CONFIG_PATH environment variable if you
installed software in a non-standard prefix.

Alternatively, you may set the environment variables poppler_CFLAGS
and poppler_LIBS to avoid the need to call pkg-config.
See the pkg-config man page for more details.
===========================
     Build failed.  ;o(    
===========================
Note: maybe try the '-d' option.

Comint exited abnormally with code 1 at Sat Aug 10 15:43:08

**Comment by vedang** [comment]
The fact that the package is being built properly means that the likely culprit is Emacs not able to setup your environment properly (i.e. Processes launched from inside Emacs don't have access to the proper environment and therefore can't find dependencies). 

I am not sure how to help you fix this or diagnose it further. 

One thing you can try (not a fix, but maybe a workaround) is to `(setq pdf-info-epdfinfo-program "/path/to/server/epdfinfo"))` before the call to `pdf-tools-install`. The path here is the same as the one in the terminal screenshot, which contains the latest `epdfinfo` executable. 

Restart Emacs and see if it can use the latest epdfinfo properly. 

The real fix is to figure out what part of your config is causing Emacs process to spawn incorrectly.

**Comment by vedang** [comment]
Consider installing and using https://github.com/purcell/exec-path-from-shell in your Emacs config

**Comment by deccanqueen88** [comment]
Hi Vedang, thanks for taking the time to address this. Since emacs is the problem, I'm wondering if it has something to do with how I have installed it. For the 27.1 version, I ran sudo apt install emacs on the terminal. When I try running sudo apt install emacs emacs-common, it tells me that emacs 27.1 is the newest version. 
For emacs 29.4, I installed it using Pop Shop, and it tells me that it installed the flatpak version. Is there some sandboxing at play here that is preventing it access?
Is there some other way to get the latest version of emacs through apt and not flatpak via the terminal, and do you think that might be a possible fix?
P.S.: I tried introducing the setq parameter as you suggested, but that did not work. I'm still trying to figure out how to use the exec-path-from-shell.

**Reply by deccanqueen88** [reply]
Hi Vedang, I'm just writing to report that the problem has been sorted for me. This time I installed the latest version of emacs using snap, and pdf-tools was able to successfully compile epdfinfo on startup. I am not sure why it was not able to do it when I downloaded Emacs 29.4 from pop shop. Anyway, thanks for the help!
